lib/SILOptimizer/Mandatory/FlowIsolation.cpp

```diff
@@ -0,0 +1,608 @@
+#include "swift/SIL/BasicBlockBits.h"
+#include "swift/SIL/DebugUtils.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+
+using namespace swift;
+
+namespace {
+
+class AnalysisInfo;
+
+// MARK: utilities
+
+static SILFunction* getCallee(SILInstruction *someInst) {
+  if (auto apply = ApplySite::isa(someInst))
+    if (SILFunction *callee = apply.getCalleeFunction())
+      return callee;
+  return nullptr;
+}
+
+// Represents the state of isolation for `self` during the flow-analysis,
+// at entry and exit to a block. The states are part of a semi-lattice,
+// where the extra top element represents a conflict in isolation:
+//
+//         T = "top"
+//        / \
+//      Iso  NonIso
+//        \ /
+//         B = "bottom"
+//
+// While we will be talking about isolated vs nonisolated uses, the only
+// isolated uses that we consider are stored property accesses.
+struct State {
+  // Each state kind, as an integer, is its position in any bit vectors.
+  enum Kind {
+    Isolated = 0,
+    Nonisolated = 1
+  };
+
+  // Number of states, excluding Top or Bottom, in this flow problem.
+  static constexpr unsigned NumStates = 2;
+};
+
+
+
+/// Information gathered for analysis that is specific to a block.
+struct Info {
+  using UseSet = SmallPtrSet<SILInstruction*, 16>;
+
+  /// Optional field to cache a non-isolated use that appears in this block,
+  /// or some predecessor block that can reach this block, is suitable for
+  /// assigning blame in a diagnostic.
+  SILInstruction* cachedBlame = nullptr;
+
+  /// Records all nonisolated uses of `self` in the block.
+  UseSet nonisolatedUses;
+
+  /// Records all stored property uses based on `self` in the block.
+  /// These are the only isolated uses that we care about.
+  UseSet propertyUses;
+
+  Info() : nonisolatedUses(), propertyUses() {}
+
+  // Diagnoses all property uses as being an error.
+  void diagnoseAll(AnalysisInfo &info, SILInstruction *blame = nullptr);
+
+  bool hasNonisolatedUse() const {
+    return !nonisolatedUses.empty();
+  }
+
+  bool hasPropertyUse() const {
+    return !propertyUses.empty();
+  }
+
+  void dump() const LLVM_ATTRIBUTE_USED {
+    llvm::dbgs() << "nonisolatedUses:\n";
+    for (auto const *i : nonisolatedUses)
+      i->dump();
+
+    llvm::dbgs() << "propertyUses:\n";
+    for (auto const *i : propertyUses)
+      i->dump();
+  }
+};
+
+/// Carries the state of analysis for an entire SILFunction.
+class AnalysisInfo : public BasicBlockData<Info> {
+private:
+  /// Isolation state at the start of the entry block to this function.
+  /// This should always be `isolated`, unless if this is a `defer`.
+  State::Kind startingIsolation = State::Isolated;
+
+public:
+
+  // The deferBlocks information is shared between all blocks of
+  // this analysis information's function.
+  llvm::SmallMapVector< SILFunction*,
+                        std::unique_ptr<AnalysisInfo>, 8> deferBlocks;
+
+  // Only computed after calling solve()
+  BitDataflow flow;
```

Comment:
This is using 4 bitvectors per basic block to analyze the uses of a single SSA value! It can all be done with just two BasicBlockSets for the entire function. One for the isolated blocks and one for the nonisolated blocks.

Reply:
Hm, yea. I might be able to implement this even more efficiently than that, too. 

I think the only information I truly need is:
1. whether a block generates nonisolation, which is already tracked with the `nonisolatedUses` set in each block's corresponding `Info`.
2. whether the block exits with nonisolation, which can be one BasicBlockSet for the whole function.

Once I've got a bunch more tests in place to make sure it's working the way we need, I'll go ahead and rewrite it to be more memory efficient, dropping the need for BitDataflow.